crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2830,6 +2830,91 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::USub, Type::BooleanLiteral(bool)) => Type::IntLiteral(-i64::from(bool)),
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
+            (UnaryOp::Not, Type::Instance(InstanceType { class })) => {
+                // TODO: move to Type::bool method
+                // I initially wrote this code in the bool method but it caused a cycle.
+                // Also this requires refactoring of that method to pass the node and diagnostics
+                // so we can emit the diagnostic when it's not possible to get the boolean of the
+                // type.
```

Comment:
Should I refactor the bool method so we can call `return_ty_result` there and move this piece of code there?

Reply:
I think with the adjustments the semantics that I outlined above in comments on the tests, this becomes much simpler and at least some of these problems disappear.

1) We don't need to emit diagnostics here, because invalid `__bool__` and `__len__` methods should cause diagnostics in the class definition, not here.

2) I think this case doesn't even require any special logic for `Type::Instance` here; we should just call `.bool()` on the type and negate it.

3) Within `Type::bool`, in the case for `Type::Instance`, I think we should be able to do `instance_ty.call_dunder("__bool__")` and check the return value to see if it is `Literal[True]` or `Literal[False]` -- for anything else, we return `Truthiness::Ambiguous`.

The only point I'm unsure of is the cycle you mentioned. It seems to me that once we simplify this code, the cycle should not be possible; `Type::call` only calls `Type::bool` if the thing being called is the `bool` builtin itself, and the `__bool__` method on a type should generally not be `builtins.bool`. It would be worth adding a test for the very weird edge case where someone does `__bool__ = bool` on a type; maybe we do need to add a special case to avoid panic here?

Reply:
With your suggestion the cycle is gone. I initially was calling `bool` on the return type of the `__bool__` and that was the reason of cycle but now I am checking the output for `Literal[False]` and `Literal[True]`.

Reply:
Yes after looking at it, I think you're right that `Type::bool` is the best place to handle this.